CRM/Contribute/BAO/Contribution.php

```diff
@@ -4506,24 +4534,32 @@ public static function completeOrder(&$input, &$ids, $objects, $transaction, $re
       }
     }
 
-    $contributionParams['id'] = $contribution->id;
-
-    $contributionResult = civicrm_api3('Contribution', 'create', $contributionParams);
-
-    // Add new soft credit against current $contribution.
-    if (CRM_Utils_Array::value('contributionRecur', $objects) && $objects['contributionRecur']->id) {
-      CRM_Contribute_BAO_ContributionRecur::addrecurSoftCredit($objects['contributionRecur']->id, $contribution->id);
-    }
-
-    $paymentProcessorId = '';
+    $paymentProcessorId = $paymentProcessorType = '';
     if (isset($objects['paymentProcessor'])) {
       if (is_array($objects['paymentProcessor'])) {
         $paymentProcessorId = $objects['paymentProcessor']['id'];
+        $paymentProcessorType = $objects['paymentProcessor']['payment_type'];
       }
       else {
         $paymentProcessorId = $objects['paymentProcessor']->id;
+        $paymentProcessorType = $objects['paymentProcessor']->payment_type;
       }
     }
+    if (empty($contributionParams['payment_instrument_id']) && $paymentProcessorType) {
+      $paymentMethod = CRM_Core_PseudoConstant::get('CRM_Contribute_DAO_Contribution', 'payment_instrument_id', array(
+        'labelColumn' => 'name',
+      ));
+      $contributionParams['payment_instrument_id'] = $paymentProcessorType == 1 ? array_search('Credit Card', $paymentMethod) : array_search('Debit Card', $paymentMethod);
```

Comment:
In 4.7 the payment_instrument_id is stored in a field against the payment processor.